compact.go

```diff
@@ -844,31 +839,107 @@ func (c *LeveledCompactor) populateBlock(blocks []BlockReader, meta *BlockMeta,
 			}
 			valset.set(l.Value)
 		}
-		postings.Add(i, lset)
-
-		i++
 	}
 	if set.Err() != nil {
 		return errors.Wrap(set.Err(), "iterate compaction set")
 	}
 
-	s := make([]string, 0, 256)
+	if meta.Stats.NumSamples == 0 {
+		// No postings to write, exit early.
+		return nil
+	}
+
+	return c.writePostings(indexw, values, set.seriesMap, indexReaders)
+}
+
+// writePostings writes the postings into the index file.
+func (c *LeveledCompactor) writePostings(indexw IndexWriter, values map[string]stringset,
+	seriesMap []map[uint64]uint64, indexReaders []IndexReader) (err error) {
+	var (
+		maxNumValues      int
+		numLabelValues    int
+		names             = make([]string, 0, len(values)+1)
+		apkName, apkValue = index.AllPostingsKey()
+	)
+	for _, v := range values {
+		numLabelValues += len(v)
+		if len(v) > maxNumValues {
+			maxNumValues = len(v)
+		}
+	}
+
+	labelValuesBuf := make([]string, 0, maxNumValues)
 	for n, v := range values {
-		s = s[:0]
+		labelValuesBuf = labelValuesBuf[:0]
+		names = append(names, n)
 
-		for x := range v {
-			s = append(s, x)
+		for val := range v {
+			labelValuesBuf = append(labelValuesBuf, val)
 		}
-		if err := indexw.WriteLabelIndex([]string{n}, s); err != nil {
+		if err := indexw.WriteLabelIndex([]string{n}, labelValuesBuf); err != nil {
 			return errors.Wrap(err, "write label index")
 		}
 	}
+	names = append(names, apkName)
+	sort.Strings(names)
+
+	if idxw, ok := indexw.(*index.Writer); ok {
+		idxw.HintPostingsWriteCount(numLabelValues)
+	}
 
-	for _, l := range postings.SortedKeys() {
-		if err := indexw.WritePostings(l.Name, l.Value, postings.Get(l.Name, l.Value)); err != nil {
-			return errors.Wrap(err, "write postings")
+	postingBuf := make([]uint64, 0, 1000000)
+	var bigEndianPost index.Postings = index.NewBigEndianPostings(nil)
+	var listPost = index.NewListPostings(nil).(*index.ListPostings)
+	for _, n := range names {
+		labelValuesBuf = labelValuesBuf[:0]
+		if n == apkName {
+			labelValuesBuf = append(labelValuesBuf, apkValue)
+		} else {
+			for v := range values[n] {
+				labelValuesBuf = append(labelValuesBuf, v)
+			}
+			sort.Strings(labelValuesBuf)
+		}
+
+		for _, v := range labelValuesBuf {
+			postingBuf = postingBuf[:0]
+			for i, ir := range indexReaders {
+				bigEndianPost, err = ir.Postings(n, v, bigEndianPost)
+				if err != nil {
+					return errors.Wrap(err, "read postings")
+				}
+				sMap := seriesMap[i]
+				idx, lastIdx := -1, -1
+				for bigEndianPost.Next() {
+					newVal, ok := sMap[bigEndianPost.At()]
```

Comment:
Oof, complexity :) I wonder if just appending everything to `postingBuf` and sorting at the end is simpler without too much change in perf.

Reply:
Or even better, new type called `remappedPostings` which takes the `sMap` and `postings`, and gives `Postings` which returns remapped values. And use those and `mergedPostings` to create the new list.

Reply:
> I wonder if just appending everything to postingBuf and sorting at the end is simpler without too much change in perf.

This would need deduplication. Also, it would mean we are storing more in the memory, upto `n` times more where `n` is the number of blocks being compacted. 

> Or even better, new type called remappedPostings which takes the sMap and postings, and gives Postings which returns remapped values. And use those and mergedPostings to create the new list.

This could make it more readable. My idea is that `remappedPostings` would take postings one by one and add it in the same way as now, and it would take `seriesMap` and not `sMap` (from the code). If we want to re-use the `mergedPostings`, it would again mean keeping postings from all the blocks together. Instead, a modified `mergedPostings` which maps the postings on the fly.